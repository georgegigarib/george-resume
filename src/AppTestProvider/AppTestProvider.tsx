import React from 'react'
import { I18nextProvider } from 'react-i18next'
import i18n from '@/locales/i18n.config'
import store from '@/store'
import { Provider } from 'react-redux'
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom'
import { vi } from 'vitest'

vi.mock('swiper/react', () => ({
  Swiper: ({
    children,
    onSlideChange,
  }: {
    children: React.ReactNode
    onSlideChange?: (params: { realIndex: number }) => void
  }) => {
    onSlideChange?.({ realIndex: 0 })
    return <div data-testid="swiper">{children}</div>
  },
  SwiperSlide: ({ children }: { children: React.ReactNode }) => <div data-testid="swiper-slide">{children}</div>,
}))

const AppTestProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <I18nextProvider i18n={i18n}>
      <Provider store={store}>
        <Router>
          <Routes>
            <Route path="/" element={children} />
          </Routes>
        </Router>
      </Provider>
    </I18nextProvider>
  )
}

export default AppTestProvider
