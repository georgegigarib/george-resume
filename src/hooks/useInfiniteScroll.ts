import { useEffect, useRef } from 'react'

const useInfiniteScroll = (containerId: string) => {
  const containerRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    const container = containerRef.current
    if (!container) return

    let animationFrame: number
    let scrollPosition = 0

    try {
      // Duplicate the content to allow infinite scrolling
      const originalContent = container.innerHTML
      container.innerHTML += originalContent

      // We calculate the total width to adjust the scrolling
      const scrollWidth = container.scrollWidth / 2

      const scrollStep = () => {
        scrollPosition += 0.5 // Adjust the scrolling speed here
        if (scrollPosition >= scrollWidth) {
          scrollPosition = 0
        }
        container.scrollLeft = scrollPosition
        animationFrame = requestAnimationFrame(scrollStep)
      }

      scrollStep()
    } catch (error) {
      console.error('Error in useInfiniteScroll:', error)
    }

    // Clean on unmount
    return () => {
      if (animationFrame) cancelAnimationFrame(animationFrame)
    }
  }, [containerId])

  return containerRef
}

export default useInfiniteScroll
