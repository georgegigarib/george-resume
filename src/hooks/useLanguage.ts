import { useTranslation } from 'react-i18next';

export const useLanguage = () => {
  const { i18n } = useTranslation();

  const changeLanguage = (lng: string) => {
    i18n.changeLanguage(lng);
    localStorage.setItem('language', lng);
  };

  const loadLanguageFromStorage = () => {
    const savedLanguage = localStorage.getItem('language') || 'en';
    i18n.changeLanguage(savedLanguage);
  };

  return { changeLanguage, loadLanguageFromStorage };
};
