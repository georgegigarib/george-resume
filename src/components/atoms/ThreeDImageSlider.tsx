import clsx from 'clsx'
import React from 'react'
import { useIsMobile } from '@/hooks/useIsMobile'
import '@/styles/slider-carousel.css'
import ImageWithLoader from '@/components/atoms/ImageLoader.tsx'

type ThreeDImageSliderProps = {
  projects: {
    src: string
    title: string
    onClick: () => void
  }[]
  isDesktopMode: boolean
}

const ThreeDImageSlider: React.FC<ThreeDImageSliderProps> = ({ projects, isDesktopMode }) => {
  const isMobile = useIsMobile()

  return (
    <div
      id="drag-container"
      className={clsx(isMobile ? '-top-[10%]' : ' top-[7vh]', isDesktopMode && isMobile ? 'top-[8%]' : '')}
    >
      <div id="spin-container">
        {projects.map((project, index) => (
          <ImageWithLoader
            key={index}
            imagePath={project.src}
            alt={project.title}
            onClick={project.onClick}
            imgClassName="carousel-image"
          />
        ))}
      </div>
      <div id="ground" />
    </div>
  )
}

export default ThreeDImageSlider
