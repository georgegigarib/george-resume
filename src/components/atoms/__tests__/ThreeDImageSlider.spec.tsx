import { render, fireEvent } from '@testing-library/react'
import { describe, expect, it, vi, beforeEach, Mock } from 'vitest'
import ThreeDImageSlider from '@/components/atoms/ThreeDImageSlider'
import AppTestProvider from '@/AppTestProvider/AppTestProvider'
import { useIsMobile } from '@/hooks/useIsMobile'

vi.mock('@/hooks/useIsMobile', () => ({
  useIsMobile: vi.fn(),
}))

describe('ThreeDImageSlider component', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const renderComponent = (isMobile: boolean, isDesktopMode: boolean, projects: any[]) => {
    ;(useIsMobile as Mock).mockReturnValue(isMobile)
    return render(
      <AppTestProvider>
        <ThreeDImageSlider projects={projects} isDesktopMode={isDesktopMode} />
      </AppTestProvider>
    )
  }

  const projects = [
    { src: 'image1.jpg', title: 'Project 1', onClick: vi.fn() },
    { src: 'image2.jpg', title: 'Project 2', onClick: vi.fn() },
  ]

  it('matches the snapshot for desktop mode', () => {
    const { asFragment } = renderComponent(false, true, projects)
    expect(asFragment()).toMatchSnapshot()
  })

  it('matches the snapshot for mobile mode', () => {
    const { asFragment } = renderComponent(true, false, projects)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders the correct number of images', () => {
    const { getAllByRole } = renderComponent(false, true, projects)
    const images = getAllByRole('img')
    expect(images.length).toBe(projects.length)
  })

  it('calls onClick when an image is clicked', () => {
    const { getAllByRole } = renderComponent(false, true, projects)
    const images = getAllByRole('img')
    fireEvent.click(images[0])
    expect(projects[0].onClick).toHaveBeenCalled()
  })
})
