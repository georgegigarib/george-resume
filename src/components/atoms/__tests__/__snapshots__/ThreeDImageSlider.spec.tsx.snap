// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThreeDImageSlider component > matches the snapshot for desktop mode 1`] = `
<DocumentFragment>
  <div
    class=" top-[7vh]"
    id="drag-container"
  >
    <div
      id="spin-container"
    >
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1yonwsg-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-1a8epmm-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
      <img
        alt="Project 1"
        class="carousel-image opacity-0"
        src="image1.jpg"
      />
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1yonwsg-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-1a8epmm-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
      <img
        alt="Project 2"
        class="carousel-image opacity-0"
        src="image2.jpg"
      />
    </div>
    <div
      id="ground"
    />
  </div>
</DocumentFragment>
`;

exports[`ThreeDImageSlider component > matches the snapshot for mobile mode 1`] = `
<DocumentFragment>
  <div
    class="-top-[10%]"
    id="drag-container"
  >
    <div
      id="spin-container"
    >
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1yonwsg-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-1a8epmm-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
      <img
        alt="Project 1"
        class="carousel-image opacity-0"
        src="image1.jpg"
      />
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1yonwsg-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-1a8epmm-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
      <img
        alt="Project 2"
        class="carousel-image opacity-0"
        src="image2.jpg"
      />
    </div>
    <div
      id="ground"
    />
  </div>
</DocumentFragment>
`;
